File: languages/js/test/parity.ts
import { Oso } from '../src/Oso';
import { Variable } from '../src/Variable';

const oso = new Oso();

class A {
  readonly x: string;

  constructor(x: string) {
    this.x = x;
  }

  foo() {
    return -1;
  }
}
oso.registerClass(A);

class D extends A {}

namespace B {
  export class C {
    readonly y: string;

    constructor(y: string) {
      this.y = y;
    }

    foo() {
      return -1;
    }
  }
}

oso.registerClass(B.C, 'C');

class E {
  static sum(args: number[]) {
    return args.reduce((a, b) => {
      return a + b;
    }, 0);
  }
}

oso.registerClass(E);

(async function () {
  // This path has the same nesting for development and the parity test jobs by sheer coincidence.
  // In tests it's `languages/js/test/parity.ts`
  // In parity tests it's `js_package/dist/test/parity.js`
  // In both these cases the relative path to the test.polar file is the same.
  const { join } = require('path');
  await oso.loadFile(join(__dirname, '../../../test/test.polar'));

  if (!(await oso.isAllowed('a', 'b', 'c'))) throw new Error();

  // Test that a built in string method can be called.
  await oso.loadStr('?= x = "hello world!" and x.endsWith("world!");');

  // Test that a custom error type is thrown.
  let exceptionThrown = false;
  try {
    await oso.loadStr('missingSemicolon()');
  } catch (e) {
    const expectedName = 'ParseError::UnrecognizedEOF';
    const expectedMessage =
      'hit the end of the file unexpectedly. Did you forget a semi-colon at line 1, column 19';
    if (e.name === expectedName && e.message === expectedMessage)
      exceptionThrown = true;
  } finally {
    if (!exceptionThrown) throw new Error();
  }

  if (
    [
      (
        await oso
          .queryRule('specializers', new D('hello'), new B.C('hello'))
          .next()
      ).done,
      (await oso.queryRule('floatLists').next()).done,
      (await oso.queryRule('intDicts').next()).done,
      (await oso.queryRule('comparisons').next()).done,
      (await oso.queryRule('testForall').next()).done,
      (await oso.queryRule('testRest').next()).done,
      (await oso.queryRule('testMatches', new A('hello')).next()).done,
      (
        await oso
          .queryRule('testMethodCalls', new A('hello'), new B.C('hello'))
          .next()
      ).done,
      (await oso.queryRule('testOr').next()).done,
      (await oso.queryRule('testUnifyClass', A).next()).done,
    ].some(v => v)
  )
    throw new Error();

  // Test that cut doesn't return anything.
  if (!(await oso.queryRule('testCut').next()).done) throw new Error();

  // Test that a constant can be called.
  oso.registerConstant(Math, 'MyMath');
  await oso.loadStr('?= MyMath.acos(1.0) = 0;');

  // test iterables work
  if ((await oso.queryRule('testIterables').next()).done) throw new Error();

  // Test built-in type specializers.
  if (
    [
      (await oso.query('builtinSpecializers(true, "Boolean")').next()).done,
      !(await oso.query('builtinSpecializers(false, "Boolean")').next()).done,
      (await oso.query('builtinSpecializers(2, "Integer")').next()).done,
      (await oso.query('builtinSpecializers(1, "Integer")').next()).done,
      !(await oso.query('builtinSpecializers(0, "Integer")').next()).done,
      !(await oso.query('builtinSpecializers(-1, "Integer")').next()).done,
      (await oso.query('builtinSpecializers(1.0, "Float")').next()).done,
      !(await oso.query('builtinSpecializers(0.0, "Float")').next()).done,
      !(await oso.query('builtinSpecializers(-1.0, "Float")').next()).done,
      (
        await oso
          .query('builtinSpecializers(["foo", "bar", "baz"], "List")')
          .next()
      ).done,
      !(
        await oso
          .query('builtinSpecializers(["bar", "foo", "baz"], "List")')
          .next()
      ).done,
      (
        await oso
          .query('builtinSpecializers({foo: "foo"}, "Dictionary")')
          .next()
      ).done,
      !(
        await oso
          .query('builtinSpecializers({foo: "bar"}, "Dictionary")')
          .next()
      ).done,
      (await oso.query('builtinSpecializers("foo", "String")').next()).done,
      !(await oso.query('builtinSpecializers("bar", "String")').next()).done,

      !(await oso.query('builtinSpecializers(1, "IntegerWithFields")').next())
        .done,
      !(
        await oso
          .query('builtinSpecializers(2, "IntegerWithGarbageFields")')
          .next()
      ).done,
      !(
        await oso
          .query('builtinSpecializers({}, "DictionaryWithFields")')
          .next()
      ).done,
      !(
        await oso
          .query('builtinSpecializers({z: 1}, "DictionaryWithFields")')
          .next()
      ).done,
      (
        await oso
          .query('builtinSpecializers({y: 1}, "DictionaryWithFields")')
          .next()
      ).done,
    ].some(v => v)
  )
    throw new Error();

  // Test deref behaviour
  await oso.loadStr(
    '?= x = 1 and E.sum([x, 2, x]) = 4 and [3, 2, x].indexOf(1) = 2;'
  );

  // Test unspecialized rule ordering
  const result = oso.queryRule(
    'testUnspecializedRuleOrder',
    'foo',
    'bar',
    new Variable('z')
  );
  if (((await result.next()).value as Map<string, any>).get('z') !== 1)
    throw new Error();
  if (((await result.next()).value as Map<string, any>).get('z') !== 2)
    throw new Error();
  if (((await result.next()).value as Map<string, any>).get('z') !== 3)
    throw new Error();

  console.log('tests pass');
})();

File: languages/rust/tests/parity.rs
use oso::host::Class;
use oso::polar::Polar;

struct A {
    x: String,
}

impl A {
    pub fn new(x: String) -> Self {
        Self { x }
    }

    pub fn foo(&self) -> i32 {
        -1
    }
}

// pub trait A {}

struct D;

// impl A for D {

// }

// oso.register_class(A)

pub mod b {
    #[derive(Default)]
    pub struct C {
        pub y: String,
    }

    impl C {
        pub fn new(y: String) -> Self {
            Self { y }
        }

        pub fn foo(&self) -> i32 {
            -1
        }
    }
}

pub fn custom_c_constructor(y: String) -> b::C {
    b::C::new(y)
}

// oso.register_class(B.C, name="C", from_polar=custom_c_constructor)

#[test]
fn test() {
    let mut polar = Polar::new();

    let mut a_class = Class::with_constructor(A::new);
    a_class.add_attribute_getter("x", |a_self: &A| a_self.x.clone());
    a_class.add_method("foo", A::foo);
    polar
        .register_class(a_class, Some("A".to_string()))
        .unwrap();

    let mut c_class = Class::with_constructor(b::C::new);
    c_class.add_attribute_getter("y", |c: &b::C| c.y.clone());
    c_class.add_method("foo", b::C::foo);
    polar
        .register_class(c_class, Some("C".to_string()))
        .unwrap();

    let polar_file = std::env::var("CARGO_MANIFEST_DIR").unwrap() + "/../../test/test.polar";
    println!("Loading: {}", polar_file);
    polar.load_file(&polar_file).unwrap();
}

// polar_file = os.path.dirname(os.path.realpath(__file__)) + "/test.polar"
// oso.load_file(polar_file)

// assert oso.is_allowed("a", "b", "c")

// # Test that a built in string method can be called.
// oso.load_str("""?= x = "hello world!" and x.endswith("world!");""")

// # Test that a custom error type is thrown.
// exception_thrown = False
// try:
//     oso.load_str("missingSemicolon()")
// except UnrecognizedEOF as e:
//     exception_thrown = True
//     assert (
//         str(e)
//         == "hit the end of the file unexpectedly. Did you forget a semi-colon at line 1, column 19"
//     )
// assert exception_thrown

// assert list(oso.query_rule("specializers", D("hello"), B.C("hello")))
// assert list(oso.query_rule("floatLists"))
// assert list(oso.query_rule("intDicts"))
// assert list(oso.query_rule("comparisons"))
// assert list(oso.query_rule("testForall"))
// assert list(oso.query_rule("testRest"))
// assert list(oso.query_rule("testMatches", A("hello")))
// assert list(oso.query_rule("testMethodCalls", A("hello"), B.C("hello")))
// assert list(oso.query_rule("testOr"))
// assert list(oso.query_rule("testHttpAndPathMapper"))

// # Test that cut doesn't return anything.
// assert not list(oso.query_rule("testCut"))

// # Test that a constant can be called.
// oso.register_constant("Math", math)
// oso.load_str("?= Math.factorial(5) == 120;")

// # Test built-in type specializers.
// assert list(oso.query('builtinSpecializers(true, "Boolean")'))
// assert not list(oso.query('builtinSpecializers(false, "Boolean")'))
// assert list(oso.query('builtinSpecializers(2, "Integer")'))
// assert list(oso.query('builtinSpecializers(1, "Integer")'))
// assert not list(oso.query('builtinSpecializers(0, "Integer")'))
// assert not list(oso.query('builtinSpecializers(-1, "Integer")'))
// assert list(oso.query('builtinSpecializers(1.0, "Float")'))
// assert not list(oso.query('builtinSpecializers(0.0, "Float")'))
// assert not list(oso.query('builtinSpecializers(-1.0, "Float")'))
// assert list(oso.query('builtinSpecializers(["foo", "bar", "baz"], "List")'))
// assert not list(oso.query('builtinSpecializers(["bar", "foo", "baz"], "List")'))
// assert list(oso.query('builtinSpecializers({foo: "foo"}, "Dictionary")'))
// assert not list(oso.query('builtinSpecializers({foo: "bar"}, "Dictionary")'))
// assert list(oso.query('builtinSpecializers("foo", "String")'))
// assert not list(oso.query('builtinSpecializers("bar", "String")'))
